assets/js/modules/analytics-4/components/widgets/MostEngagingPagesWidget.js

```diff
@@ -0,0 +1,240 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import PropTypes from 'prop-types';
+
+/**
+ * WordPress dependencies
+ */
+import { createInterpolateElement } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import Data from 'googlesitekit-data';
+import { CORE_USER } from '../../../../googlesitekit/datastore/user/constants';
+import {
+	DATE_RANGE_OFFSET,
+	MODULES_ANALYTICS_4,
+} from '../../datastore/constants';
+import {
+	MetricTileTable,
+	MetricTileTablePlainText,
+} from '../../../../components/KeyMetrics';
+import Link from '../../../../components/Link';
+import { ZeroDataMessage } from '../../../analytics/components/common';
+import whenActive from '../../../../util/when-active';
+import ConnectGA4CTATileWidget from './ConnectGA4CTATileWidget';
+import useViewOnly from '../../../../hooks/useViewOnly';
+import { numFmt } from '../../../../util';
+const { useSelect, useInViewSelect } = Data;
+
+function MostEngagingPagesWidget( props ) {
+	const { Widget } = props;
+
+	const viewOnlyDashboard = useViewOnly();
+
+	const dates = useSelect( ( select ) =>
+		select( CORE_USER ).getDateRangeDates( {
+			offsetDays: DATE_RANGE_OFFSET,
+		} )
+	);
+
+	const pageViewsReportOptions = {
+		...dates,
+		dimensions: [ 'pagePath' ],
+		metrics: [ { name: 'screenPageViews' } ],
+	};
+
+	const pageViewsReport = useInViewSelect( ( select ) =>
+		select( MODULES_ANALYTICS_4 ).getReport( pageViewsReportOptions )
+	);
+
+	const averagePageViews =
+		Math.round(
+			pageViewsReport?.totals?.[ 0 ]?.metricValues[ 0 ].value /
+				pageViewsReport?.rowCount
+		) || 0;
+
+	const reportOptions = {
+		...dates,
+		dimensions: [ 'pagePath' ],
+		metrics: [ 'engagementRate', 'screenPageViews' ],
+		orderby: [
+			{
+				metric: { metricName: 'engagementRate' },
+				desc: true,
+			},
+			{
+				metric: { metricName: 'screenPageViews' },
+				desc: true,
+			},
+		],
+		metricFilters: {
+			screenPageViews: {
+				filterType: 'numericFilter',
+				operation: 'GREATER_THAN_OR_EQUAL',
+				value: { int64Value: averagePageViews },
+			},
+		},
+		limit: 3,
+	};
+
+	const pageViewsReportErrors = useSelect( ( select ) =>
+		select( MODULES_ANALYTICS_4 ).getErrorForSelector( 'getReport', [
+			pageViewsReportOptions,
+		] )
+	);
+
+	const error = useSelect( ( select ) => {
+		const reportOptionsErrors = select(
+			MODULES_ANALYTICS_4
+		).getErrorForSelector( 'getReport', [ reportOptions ] );
+
+		if ( pageViewsReportErrors && reportOptionsErrors ) {
+			return [ pageViewsReportErrors, reportOptionsErrors ];
+		}
+
+		return pageViewsReportErrors || reportOptionsErrors || undefined;
+	} );
+
+	const hasFinishedResolvingPageViewReport = useSelect( ( select ) =>
+		select( MODULES_ANALYTICS_4 ).hasFinishedResolution( 'getReport', [
+			pageViewsReportOptions,
+		] )
+	);
+
+	const report = useSelect( ( select ) => {
```

Comment:
We should use the `useInViewSelect` hook to load this report only when this widget gets into the viewport.